indexing-service/src/main/java/org/apache/druid/indexing/common/actions/RetrieveUpgradedToSegmentsIdsAction.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.actions;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.ImmutableList;
+import org.apache.druid.indexing.common.task.Task;
+import org.apache.druid.metadata.SegmentUpgradeInfo;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Task action to retrieve the segment upgrade infos of a given set of used or unused upgraded segment ids.
+ */
+public class RetrieveUpgradedToSegmentsIdsAction implements TaskAction<List<SegmentUpgradeInfo>>
+{
+  private final String dataSource;
+  private final Set<String> upgradedSegmentIds;
+
+  @JsonCreator
+  public RetrieveUpgradedToSegmentsIdsAction(
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("upgradedSegmentIds") Set<String> upgradedSegmentIds
+  )
+  {
+    this.dataSource = dataSource;
+    this.upgradedSegmentIds = upgradedSegmentIds;
+  }
+
+  @JsonProperty
+  public String getDataSource()
+  {
+    return dataSource;
+  }
+
+  @JsonProperty
+  public Set<String> getUpgradedSegmentIds()
+  {
+    return upgradedSegmentIds;
+  }
+
+  @Override
+  public TypeReference<List<SegmentUpgradeInfo>> getReturnTypeReference()
+  {
+    return new TypeReference<List<SegmentUpgradeInfo>>()
+    {
+    };
+  }
+
+  @Override
+  public List<SegmentUpgradeInfo> perform(Task task, TaskActionToolbox toolbox)
```

Comment:
The problem is the structure of the payload rather than its size.

A list of `SegmentUpdateInfo` makes sense for the other action `RetrieveUpgradedFromSegmentIdsAction`, because you would have exactly one entry in the output for every input segment.
But here, it feels weird since you would have multiple entries in the list for the same parent ID.

The action is asking for all the children of a parent. So the response has to answer that appropriately.

>  It also allows us to add other fields like a load spec or a new column in the future if needed without having to introduce a new task action.

If you want to keep that flexibility, you can create a class that contains the whole map inside it.

The better alternatives are these:
1. Just return a List of segment IDs that are children of any of the input segment IDs.
2. Return a Map as suggested above. This is better than the list as the relationships are clear.
3. Create a separate class `SegmentUpgradeMappings` or something and put the _whole_ map inside it. The only advantage of this is that we might be able to extend it further in the future.
4. If you want to the `SegmentUpgradeInfo` thing, the right way to do it would be as follows:
  a. Rename `SegmentUpgradeInfo` to `SegmentUpgradeMapping`
  b. `SegmentUpgradeMapping` should have a single `segmentId` and a list `upgradedToSegmentIds`.
  c. Keep only one entry in the returned list for every input ID.
  d. If you look at this closely though, it is nothing but a map with extra steps.

Reply:
Followed option 3